In [1521]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np

# Part 1: Load the dataset

In [1522]:
# Load the given datasets
numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")
categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")

# Print the data

display(numerical)
display(categorical)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

The data in the datasets can be used to help predict if a patient will have chronic kidney disease. The features of the dataset are all the different variables that help determine if someone has chronic kidney disorder, and the label is the variable "Target", which is whether or not someone has ckd.

All the features in the numerical dataset are numerical features (such as age, bp, and bgr). All the features in the categorical dataset are either categorical (such as rbc, pc, and bs), or nominal (such as al and su).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [1523]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_before = len(numerical)
num_no_dup = numerical.drop_duplicates(subset="unique_id")
num_after = len(num_no_dup)
if num_before > num_after:
    num_diff = num_before - num_after
    print("Duplicates in numerical:", num_diff)

# Drop these duplicate rows


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cat_before = len(categorical)
cat_no_dup = categorical.drop_duplicates(subset="unique_id")
cat_after = len(cat_no_dup)
if cat_before > cat_after:
    cat_diff = cat_before - cat_after
    print("Duplicates in categorical:", cat_diff)


Duplicates in numerical: 30
Duplicates in categorical: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:
All the unique ids should be in both datasets. Each unique id represents a different patient, and if they aren't in both datasets, then you have some data missing on them. If they are missing from one of the datasets, then you are only going to have half of the information on them.

In [1524]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
df = pd.merge(num_no_dup, cat_no_dup)
#Print the combined dataset
display(df)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5,0.0,0.0,NaN,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
385,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0,1.0,0.0,NaN,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4,4.0,2.0,abnormal,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
388,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,normal,abnormal,present,present,no,no,no,good,no,yes,ckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [1525]:
# Calculate the percentage of rows that contain atleast one missing value
before = len(df)
df = df.dropna(axis=0, how='any')
after = len(df)
# Print %
diff = (after/before)*100
print("Percent of rows with at least one missing value:", diff,"%")
# Drop these rows from the dataset

# Print the Dataset
display(df)

Percent of rows with at least one missing value: 39.23076923076923 %


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,8100.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
13,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
379,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
381,462892,36.0,80.0,85.0,16.0,1.1,142.0,4.1,15.6,44.0,5800.0,6.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4,4.0,2.0,abnormal,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.4: Sort the dataset according to the Labels

In [1526]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
df = df.sort_values(by=['Target'])
df = df.reset_index(drop=True)
# Print the dataset
display(df)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,normal,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,7000.0,3.2,3.0,2.0,abnormal,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,12100.0,3.7,2.0,0.0,abnormal,abnormal,present,notpresent,yes,yes,no,poor,no,yes,ckd
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,10500.0,4.1,2.0,0.0,abnormal,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,7000.0,5.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,7000.0,5.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,6200.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,6300.0,5.9,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [1527]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
df_dummies = pd.get_dummies(df, columns=categorical_columns)
# Print the dataset
display(df_dummies)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,False,True,True,False,False,True,False,True,True,False,True,False,True,False,True,False,True,False,False,True,True,False
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,7000.0,3.2,3.0,2.0,True,False,True,False,False,True,True,False,False,True,False,True,False,True,False,True,True,False,True,False,True,False
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,12100.0,3.7,2.0,0.0,True,False,True,False,False,True,True,False,False,True,False,True,True,False,False,True,True,False,False,True,True,False
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,True,False,True,False,True,False,True,False,False,True,False,True,False,True,False,True,False,True,False,True,True,False
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,10500.0,4.1,2.0,0.0,True,False,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,7000.0,5.3,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,7000.0,5.2,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,6200.0,6.2,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,6300.0,5.9,0.0,0.0,False,True,False,True,True,False,True,False,True,False,True,False,True,False,True,False,True,False,True,False,False,True


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: If there are only two possibilities for values for a column, it would be possible to use a singular column with a binary variable instead. This is not the case for categorical variables with more than two values, as a binary value is a maximum of two, and won't be able to represent all possible values.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [1528]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
temp_df = df.loc[:, numerical_columns]
lim = np.abs((temp_df - temp_df.mean()) / temp_df.std(ddof=0)) < 3
df.loc[:, numerical_columns] = temp_df.where(lim, np.nan)
df = df.dropna(axis=0, how='any')
df = df.reset_index(drop=True)

# Print the dataset
display(df)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,normal,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,7000.0,3.2,3.0,2.0,abnormal,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
3,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,10500.0,4.1,2.0,0.0,abnormal,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
4,546225,46.0,60.0,163.0,92.0,3.3,141.0,4.0,9.8,28.0,14600.0,3.2,1.0,0.0,normal,normal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,7000.0,5.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
130,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,7000.0,5.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
131,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,6200.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
132,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,6300.0,5.9,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [1529]:
# Normalize the all Numerical Attributes in the dataset.
df_min_max_scaled = df
for column in numerical_columns:
    df_min_max_scaled[column] = (df_min_max_scaled[column] - df_min_max_scaled[column].min()) / (df_min_max_scaled[column].max() - df_min_max_scaled[column].min())
df = df_min_max_scaled
# Print the dataset
display(df)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.171429,0.221311,0.184211,0.653226,0.333333,4.0,0.0,normal,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.295082,0.368421,0.217742,0.153846,3.0,2.0,abnormal,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,343710,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,0.210526,0.395161,0.153846,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
3,532520,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.657143,0.450820,0.447368,0.500000,0.384615,2.0,0.0,abnormal,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
4,546225,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.314286,0.344262,0.315789,0.830645,0.153846,1.0,0.0,normal,normal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,647552,0.743243,0.25,0.214592,0.124183,0.037037,0.566667,0.171429,0.967213,0.947368,0.217742,0.692308,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
130,125587,0.418919,0.00,0.167382,0.241830,0.086420,0.700000,0.571429,0.770492,0.842105,0.217742,0.666667,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
131,614376,0.689189,0.50,0.210300,0.045752,0.098765,0.500000,0.514286,0.803279,0.684211,0.153226,0.923077,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
132,853426,0.472973,0.25,0.236052,0.183007,0.024691,0.666667,0.600000,0.918033,0.657895,0.161290,0.846154,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Specific gravity is already removed, but it doesn't relate to either the patient's symptoms or any lab results. Appetite is a subjective term, and isn't super clear, so as a result is isn't really relevant in terms of predicting CKD in patients.

In [1530]:
#Remove that column
df = df.drop(axis=1, columns=['appet'])
# Print the dataset
df

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,pe,ane,Target
0,474407,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.171429,0.221311,0.184211,0.653226,0.333333,4.0,0.0,normal,abnormal,present,present,no,no,no,no,yes,ckd
1,137148,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.295082,0.368421,0.217742,0.153846,3.0,2.0,abnormal,abnormal,present,notpresent,yes,yes,yes,no,no,ckd
2,343710,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,0.210526,0.395161,0.153846,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,yes,yes,ckd
3,532520,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.657143,0.450820,0.447368,0.500000,0.384615,2.0,0.0,abnormal,normal,notpresent,notpresent,no,no,no,no,no,ckd
4,546225,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.314286,0.344262,0.315789,0.830645,0.153846,1.0,0.0,normal,normal,notpresent,notpresent,yes,yes,no,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,647552,0.743243,0.25,0.214592,0.124183,0.037037,0.566667,0.171429,0.967213,0.947368,0.217742,0.692308,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,no,no,notckd
130,125587,0.418919,0.00,0.167382,0.241830,0.086420,0.700000,0.571429,0.770492,0.842105,0.217742,0.666667,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,no,no,notckd
131,614376,0.689189,0.50,0.210300,0.045752,0.098765,0.500000,0.514286,0.803279,0.684211,0.153226,0.923077,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,no,no,notckd
132,853426,0.472973,0.25,0.236052,0.183007,0.024691,0.666667,0.600000,0.918033,0.657895,0.161290,0.846154,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,no,no,notckd


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [1531]:
# Export the dataframe to a new csv file
df.to_csv("chronic_kidney_disease_cleaned.csv", index=False)
# Export the dataframe to a new json file
df.to_json("chronic_kidney_disease_cleaned.json", orient='index', indent=1)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [1532]:
first_fifteen = df.head(15)
first_fifteen.to_json("first_fifteen.json", orient='index')
first_fifteen

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,pe,ane,Target
0,474407,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.171429,0.221311,0.184211,0.653226,0.333333,4.0,0.0,normal,abnormal,present,present,no,no,no,no,yes,ckd
1,137148,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.295082,0.368421,0.217742,0.153846,3.0,2.0,abnormal,abnormal,present,notpresent,yes,yes,yes,no,no,ckd
2,343710,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,0.210526,0.395161,0.153846,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,yes,yes,ckd
3,532520,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.657143,0.450820,0.447368,0.500000,0.384615,2.0,0.0,abnormal,normal,notpresent,notpresent,no,no,no,no,no,ckd
4,546225,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.314286,0.344262,0.315789,0.830645,0.153846,1.0,0.0,normal,normal,notpresent,notpresent,yes,yes,no,no,no,ckd
5,514721,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.542857,0.286885,0.342105,0.169355,0.205128,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,no,no,ckd
6,621332,0.716216,0.50,1.000000,0.163399,0.111111,0.066667,0.171429,0.393443,0.500000,0.532258,0.435897,1.0,0.0,abnormal,normal,notpresent,notpresent,no,yes,no,no,no,ckd
7,481293,0.729730,0.00,0.935622,0.169935,0.160494,0.333333,0.028571,0.188525,0.236842,0.879032,0.102564,3.0,1.0,normal,abnormal,present,notpresent,yes,no,no,no,yes,ckd
8,297893,0.837838,0.50,0.373391,0.522876,0.456790,0.333333,1.000000,0.000000,0.000000,0.540323,0.000000,3.0,2.0,normal,abnormal,present,present,yes,yes,yes,yes,no,ckd
9,615697,0.000000,0.00,0.103004,0.372549,0.074074,0.500000,0.571429,0.352459,0.368421,1.000000,0.564103,4.0,0.0,abnormal,abnormal,notpresent,present,no,no,no,no,no,ckd


### Paste the markdown table here

|unique_id| age         | bp   | bgr         | bu          | sc          | sod         | pot         | hemo        | pcv         | wbcc        | rbcc        | al | su | rbc      | pc       | pcc         | ba         | htn | dm  | cad | pe  | ane | Target |
|-----------|-------------|------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|----|----|----------|----------|-------------|------------|-----|-----|-----|-----|-----|--------|
|474407| 0.2027|0.75|0.1588| 0.1961      | 0.1605      | 0.1667      | 0.1714      | 0.2213      | 0.1842      | 0.6532      | 0.3333      | 4  | 0  | normal   | abnormal | present     | present    | no  | no  | no  | no  | yes | ckd    |
| 137148    | 0.9054     | 1.0  | 0.9657      | 0.5229      | 0.6420      | 0.6667      | 0.0         | 0.2951      | 0.3684      | 0.2177      | 0.1538      | 3  | 2  | abnormal | abnormal | present     | notpresent | yes | yes | yes | no  | no  | ckd    |
| 343710    | 0.7432     | 0.5  | 0.4421      | 0.9020      | 0.4321      | 0.5         | 0.6571      | 0.1721      | 0.2105      | 0.3952      | 0.1538      | 2  | 0  | abnormal | abnormal | notpresent  | notpresent | yes | yes | yes | yes | yes | ckd    |
| 532520    | 0.7297     | 0.75 | 0.1502      | 0.2810      | 0.2346      | 0.5333      | 0.6571      | 0.4508      | 0.4474      | 0.5         | 0.3846      | 2  | 0  | abnormal | normal   | notpresent  | notpresent | no  | no  | no  | no  | no  | ckd    |
| 546225    | 0.5405     | 0.0  | 0.3991      | 0.5359      | 0.3580      | 0.7         | 0.3143      | 0.3443      | 0.3158      | 0.8306      | 0.1538      | 1  | 0  | normal   | normal   | notpresent  | notpresent | yes | yes | no  | no  | no  | ckd    |
| 514721    | 0.6757     | 0.75 | 0.2532      | 0.6340      | 0.7778      | 0.3667      | 0.5429      | 0.2869      | 0.3421      | 0.1694      | 0.2051      | 2  | 0  | abnormal | abnormal | notpresent  | notpresent | yes | no  | no  | no  | no  | ckd    |
| 621332    | 0.7162     | 0.5  | 1.0         | 0.1634      | 0.1111      | 0.0667      | 0.1714      | 0.3934      | 0.5         | 0.5323      | 0.4359      | 1  | 0  | abnormal | normal   | notpresent  | notpresent | no  | yes | no  | no  | no  | ckd    |
| 481293    | 0.7297     | 0.0  | 0.9356      | 0.1699      | 0.1605      | 0.3333      | 0.0286      | 0.1885      | 0.2368      | 0.8790      | 0.1026      | 3  | 1  | normal   | abnormal | present     | notpresent | yes | no  | no  | no  | yes | ckd    |
| 297893    | 0.8378     | 0.5  | 0.3734      | 0.5229      | 0.4568      | 0.3333      | 1.0         | 0.0         | 0.0         | 0.5403      | 0.0         | 3  | 2  | normal   | abnormal | present     | present    | yes | yes | yes | yes | no  | ckd    |
| 615697    | 0.0       | 0.0  | 0.1030      | 0.3725      | 0.0741      | 0.5         | 0.5714      | 0.3525      | 0.3684      | 1.0         | 0.5641      | 4  | 0  | abnormal | abnormal | notpresent  | present    | no  | no  | no  | no  | no  | ckd    |
| 843362    | 0.8784     | 0.0  | 0.2060      | 0.7516      | 0.6049      | 0.5333      | 0.5714      | 0.4754      | 0.5         | 0.8790      | 0.4359      | 4  | 0  | normal   | normal   | notpresent  | notpresent | yes | yes | no  | yes | no  | ckd    |
| 995177    | 0.8514     | 0.25 | 0.6180      | 0.5621      | 0.7284      | 0.0         | 0.2857      | 0.3115      | 0.3158      | 0.5806      | 0.1795      | 4  | 3  | normal   | abnormal | present     | present    | yes | yes | yes | yes | yes | ckd    |
| 828592    | 0.8784     | 0.25 | 0.6395      | 0.4706      | 0.3951      | 0.4333      | 0.4286      | 0.3934      | 0.4474      | 0.1048      | 0.2564      | 3  | 0  | normal   | abnormal | present     | present    | yes | yes | yes | no  | no  | ckd    |
| 923613    | 0.7838     | 0.0  | 0.7253      | 0.3137      | 0.4815      | 0.5667      | 0.7143      | 0.3197      | 0.3421      | 0.2581      | 0.2051      | 4  | 1  | abnormal | abnormal | notpresent  | present    | yes | yes | no  | yes | no  | ckd    |
| 955830    | 0.6622     | 0.5  | 0.6180      | 0.4118      | 0.4321      | 0.5667      | 0.5714      | 0.4344      | 0.4737      | 0.25        | 0.2821      | 3  | 1  | normal   | abnormal | present     | present    | yes | yes | no  | yes | no  | ckd    |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

One way to mitigate data conversion hallucinations would be to convert one column at a time to reduce the amount of data being converted at once.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [1533]:
# Code to rename all the columns in the dataset

# Create a mapping dictionary for column name conversions
column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia'
}

# Rename the columns using the mapping
df.rename(columns=column_mapping, inplace=True)

# Display the updated DataFrame
df

,unique_id,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,packed cell volume,wbcc,rbcc,albumin,sugar,red blood cells,pus cell,pus cell clumps,bacteria,hypertension,diabetes mellitus,coronary artery disease,pedal edema,anemia,Target
0,474407,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.171429,0.221311,0.184211,0.653226,0.333333,4.0,0.0,normal,abnormal,present,present,no,no,no,no,yes,ckd
1,137148,0.905405,1.00,0.965665,0.522876,0.641975,0.666667,0.000000,0.295082,0.368421,0.217742,0.153846,3.0,2.0,abnormal,abnormal,present,notpresent,yes,yes,yes,no,no,ckd
2,343710,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,0.210526,0.395161,0.153846,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,yes,yes,ckd
3,532520,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.657143,0.450820,0.447368,0.500000,0.384615,2.0,0.0,abnormal,normal,notpresent,notpresent,no,no,no,no,no,ckd
4,546225,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.314286,0.344262,0.315789,0.830645,0.153846,1.0,0.0,normal,normal,notpresent,notpresent,yes,yes,no,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,647552,0.743243,0.25,0.214592,0.124183,0.037037,0.566667,0.171429,0.967213,0.947368,0.217742,0.692308,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,no,no,notckd
130,125587,0.418919,0.00,0.167382,0.241830,0.086420,0.700000,0.571429,0.770492,0.842105,0.217742,0.666667,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,no,no,notckd
131,614376,0.689189,0.50,0.210300,0.045752,0.098765,0.500000,0.514286,0.803279,0.684211,0.153226,0.923077,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,no,no,notckd
132,853426,0.472973,0.25,0.236052,0.183007,0.024691,0.666667,0.600000,0.918033,0.657895,0.161290,0.846154,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,no,no,notckd


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [1534]:
# Converted SQL to Pandas code
result = df.groupby('Target').count().reset_index()
result

,Target,unique_id,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,packed cell volume,wbcc,rbcc,albumin,sugar,red blood cells,pus cell,pus cell clumps,bacteria,hypertension,diabetes mellitus,coronary artery disease,pedal edema,anemia
0,ckd,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23,23
1,notckd,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111


*Describe what the above code does here*

The above code converts the dataframe from each row representing a different patient to a new dataframe that represents the distribution of each feature between patients with CKD and patients without CKD.